consensus/src/chained_bft/persistent_storage.rs

```diff
@@ -68,31 +115,38 @@ pub struct RecoveryData<T> {
 impl<T: Payload> RecoveryData<T> {
     pub fn new(
         last_vote: Option<Vote>,
+        ledger_recovery_data: LedgerRecoveryData<T>,
         mut blocks: Vec<Block<T>>,
         mut quorum_certs: Vec<QuorumCert>,
         storage_ledger: &LedgerInfo,
         root_executed_trees: ExecutedTrees,
         highest_timeout_certificate: Option<TimeoutCertificate>,
-        validator_keys: ValidatorSet,
     ) -> Result<Self> {
-        let root =
-            Self::find_root(&mut blocks, &mut quorum_certs, storage_ledger).with_context(|| {
-                // for better readability
-                quorum_certs.sort_by_key(|qc| qc.certified_block().round());
-                format!(
-                    "Blocks in db: {}\nQuorum Certs in db: {}\n",
-                    blocks
-                        .iter()
-                        .map(|b| format!("\n\t{}", b))
-                        .collect::<Vec<String>>()
-                        .concat(),
-                    quorum_certs
-                        .iter()
-                        .map(|qc| format!("\n\t{}", qc))
-                        .collect::<Vec<String>>()
-                        .concat(),
-                )
-            })?;
+        let genesis_root = ledger_recovery_data.genesis_root();
+        let validators = ledger_recovery_data.validators();
+        let validator_keys = ledger_recovery_data.validator_keys();
+        let root = match genesis_root {
+            Some(root) => root,
+            None => Self::find_root(&mut blocks, &mut quorum_certs, storage_ledger).with_context(
```

Comment:
do we need to move the genesis case out of `find_root`, it should work with empty blocks/qc

Reply:
Currently we extract data from ledger and consensus db, combine them and try to construct `initial_data` from it. Since we are separating these two piece of data, I think it makes more sense to construct `genesis_block` in `LedgerRecoveryData` because it's where the data used to construct `genesis_block` lives.